rfd/0179-api-versioning.md

```diff
@@ -0,0 +1,392 @@
+- What part of the feature may not work.
+- What version their proxy must be updated to in order to be fully operational. 
+
+An example of this in recent memory is Pinned Resources. The "pinned resources"
+tab was available in clients that _may_ have been communicating with proxies
+older than the required version. So a check was made to see if that proxy
+supported pinned resources, and if it wasn't, made the error known to the user
+that "Pinned Resources is supported by proxies version 14.1+" (or something like
+that).
+
+### Updating request fields in the web client
+Updating what fields are sent in a request body is ok as long as the API follows the updated request fields guidelines. 
+
+---
+### A short checklist when updating the web client/api
+- [ ] If a new endpoint was added to accommodate a new feature, the new feature must fail gracefully with the required proxy version
+- [ ] When updating a feature, check if the web client works as expected with a proxy version N-1
+- [ ] When updating a feature, check that the proxy works receiving requests from web clients N-1
+- [ ] If deprecating an endpoint, a TODO is added to delete in the relevant version
+- [ ] When updating an API response, deprecated fields are still populated
+
+---
+
+## Solution
+
+#### Incrementing version prefix
+We should [stop stripping the api prefix](https://github.com/gravitational/teleport/blob/master/lib/web/apiserver.go#L527-L530), and add a new endpoint with a incremented prefix to accommodate the new response shape.
+```go
+
+// regex to match /v1, /v2, /v3, etc
+apiPrefixRegex := regexp.MustCompile(`^/v(\d+)/webapi`)
+// request is going to the API?
+if matches := apiPrefixRegex.FindStringSubmatch(r.URL.Path); matches != nil {
+    version := matches[1]
+    // Strip the prefix only if it's v1
+    if version == "1" {
+        newPath := strings.TrimPrefix(r.URL.Path, "/v1")
+        r.URL.Path = newPath
+    }
+    h.ServeHTTP(w, r)
+}
+
+```
+
+This would allow us to follow the guidelines above when it is necessary to
+change required request/response shapes. This solves the problem of fixing
+backward compatible changes and is 'good enough'. But if we are in the mode of
+updating our DX, I propose updating our web client/server communication all
+together with ConnectRPC to not only receive help with backward compatibly with
+protomessages, but also give us type safety in the server/client code
+automatically.
+
+### Add a new RPC service for the web API
+
+**Summary**: Leverage ConnectRPC and create a web service that can use a proto
+schema to define our requests and response types. Also, optionally, use a modern
+fetching framework to simplify and unify the way we make api requests in the web
+client. This will remove boilerplate that we need to use our APIs in the web
+client, provide type safety out of the box for both server and client code, and
+provide backward compatibility guardrails when creating/updating our APIs.
```

Comment:
I totally agree with @bl-nero's comment above about bumping a version number being good enough for one-off breaking changes and the biggest advantage of Connect RPC being not forwards/backwards compatibility, but autogenerated type-checked structures in both JS and Go.

As far as compatibility is concerned, I feel like Connect RPC itself doesn't provide anything which we couldn't replicate with something like, idk, zod on the JS side. The difference is that with zod we would need to replicate the same structures in JS and then in Go.

I don't think we'll find anyone who, after reading this RFD, would be able to conclude with "This is a great that will absolutely work and be worth the time, let's do it". Perhaps it would be worthwhile to do a spike with an example implementation to see:

* How easy it'll be to integrate this along the current web API.
* How Connect RPC would address common pain points or past problems compared to the existing solution.

---

Another idea to toy with which could turn out to be an argument in favor of RPCs: how many web API endpoints do actual work and how many of them merely send off data to the auth service? Given the unusual compatibility guarantees put upon the Web UI (a client that needs to work with older servers), how feasible would it be to reuse protos from the `/api` directory to streamline endpoints which merely pass data to the auth service?

By the last question, I mean that when I want to add a new field to a usage event in Connect, which passes from the Electron app through tshd to prehog, it's a matter of updating a single proto file and the change trickles down to every place that uses it. Is this something achievable in the Web UI as well?

Reply:
> The difference is that with zod we would need to replicate the same structures in JS and then in Go.

This is the crux of the problem. At the end of the day, if we rely on creating code on the backend _and_ the frontend, we are bound to just end up in the same place we are. Removing the need to create code in two places and removing boilerplate seems worthwhile. A spike is probably a great idea (more than I spike i did for implementing a test feature, more like spike an actual feature instead). 

>  how feasible would it be to reuse protos from the /api directory to streamline endpoints which merely pass data to the auth service?

Probably pretty feasible. Although, from my experience, we send some fields that _arent_ passed to auth (and are used for creating the auth message generally) but its the return trip where things are quite different. Almost every resource-based web endpoint receives data from auth an has to convert it to a UI structure so reusing protos from proxy<->auth doesn't really seem to work.
I think the benefit that Connect gives us is the type generation is a bit more modern in that it reads closer to how types would be generated by a human, and also (if we went the tanstack query route), has built in generators for client queries as well. A lot it the type system _could_ be solved without it but from my initial testing, this was the best DX imo. 

I agree that solving the backward compatibility problem is as simple as appending the version prefix I talked about in the RFD. If we want to discuss the type system in a greater context, we can move it to its own RFD.

Reply:
> If we want to discuss the type system in a greater context, we can move it to its own RFD.

+1